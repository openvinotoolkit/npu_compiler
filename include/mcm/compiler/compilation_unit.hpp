#ifndef MV_COMPILATION_UNIT_HPP_
#define MV_COMPILATION_UNIT_HPP_

#include "include/mcm/computation/model/control_model.hpp"
#include "include/mcm/computation/model/data_model.hpp"
#include "include/mcm/op_model.hpp"
#include "include/mcm/logger/log_sender.hpp"
#include "include/mcm/pass/pass_manager.hpp"
#include "include/mcm/utils/env_loader.hpp"
#include "include/mcm/compiler/compilation_descriptor.hpp"
#include "include/mcm/compiler/compilation_profiler.hpp"
#include "include/mcm/target/kmb/runtime_model/runtime_model.hpp"

namespace mv
{

    class CompilationUnit : public LogSender
    {

        static const std::string ma2490DefTargetDescPath_;

        static const std::string ma2490DefCompDescPath_;
        static const std::string compositionalModelRecordingsPath_;

        static Logger& logger_;

        OpModel* model_;
        PassManager passManager_;
        TargetDescriptor targetDescriptor_;
        CompilationDescriptor compDescriptor_;
        const static unsigned jsonParserBufferLength_ = 256;

    public:
        std::shared_ptr<std::vector<char> > getBlob() const;

        CompilationUnit(const std::string& modelName);
        ~CompilationUnit();

        bool loadTargetDescriptor(const std::string& path);
        bool loadTargetDescriptor(Target target);
        bool loadCompilationDescriptor(const std::string& path);
        bool loadCompilationDescriptor(Target target);

        CompilationDescriptor& compilationDescriptor();
        OpModel& model();

        bool initialize();
        Element runStep();
        Element run();
        bool completed() const;

        virtual std::string getLogID() const override;

    };

}

#endif // MV_COMPILATION_UNIT_HPP_
