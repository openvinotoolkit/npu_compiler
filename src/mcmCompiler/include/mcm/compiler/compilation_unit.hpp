#ifndef MV_COMPILATION_UNIT_HPP_
#define MV_COMPILATION_UNIT_HPP_

#include "include/mcm/computation/model/control_model.hpp"
#include "include/mcm/computation/model/data_model.hpp"
#include "include/mcm/op_model.hpp"
#include "include/mcm/logger/log_sender.hpp"
#include "include/mcm/pass/pass_manager.hpp"
#include "include/mcm/utils/env_loader.hpp"
#include "include/mcm/compiler/compilation_descriptor.hpp"
#include "include/mcm/compiler/compilation_profiler.hpp"
#include "include/mcm/target/kmb/runtime_model/runtime_model.hpp"

namespace mv
{

    class CompilationUnit : public LogSender
    {

        static const std::string ma2490DefTargetDescPath_;
        static const std::string ma3100DefTargetDescPath_;
        static const std::string ma3600DefTargetDescPath_;

        static const std::string ma2490DefCompDescPath_;
        static const std::string ma3100DefCompDescPath_;
        static const std::string ma3600DefCompDescPath_;
        static const std::string compositionalModelRecordingsPath_;
        static const unsigned jsonParserBufferLength_ = 256;

        static Logger& logger_;
        static bool init_;

        OpModel* model_;
        PassManager passManager_;
        TargetDescriptor targetDescriptor_;
        CompilationDescriptor compDescriptor_;
        bool preCompiled_;

    public:

        CompilationUnit(const std::string& modelName);
        CompilationUnit(const char *blobBuffer, unsigned blobSize, TargetDescriptor td);
        CompilationUnit(const CompilationUnit & rhs) = delete;
        ~CompilationUnit();

        CompilationUnit & operator=(const CompilationUnit & rhs) = delete;
        bool loadTargetDescriptor(const std::string& path);
        bool loadTargetDescriptor(Target target);
        bool loadCompilationDescriptor(const std::string& path);
        bool loadCompilationDescriptor(Target target);

        CompilationDescriptor& compilationDescriptor();
        OpModel& model();

        bool initialize();
        Element runStep();
        Element run();
        bool completed() const;
        void reset();

        void getName(char* name, unsigned bufferSize) const;
        std::shared_ptr<std::vector<char> > getBlob() const;
        const BufferMap& getBufferMap() const;

        void loadBlob(const char *blobBuffer, unsigned blobSize);

        virtual std::string getLogID() const override;

    };

}

#endif // MV_COMPILATION_UNIT_HPP_
