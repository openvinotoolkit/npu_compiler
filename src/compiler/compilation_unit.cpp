#include "include/mcm/compiler/compilation_unit.hpp"
#include "emu/manager.hpp"

const std::string mv::CompilationUnit::ma2480DefTargetDescPath_ = "/config/target/ma2480.json";
const std::string mv::CompilationUnit::ma2490DefTargetDescPath_ = "/config/target/release_kmb.json";
const std::string mv::CompilationUnit::compositionalModelRecordingsPath_ = "/recordings/";
const std::string mv::CompilationUnit::ma2480DefCompDescPath_ = "/config/compilation/release_ma2480.json";
const std::string mv::CompilationUnit::ma2490DefCompDescPath_ = "/config/compilation/release_kmb.json";
const std::string mv::CompilationUnit::ma2490EmulatorCompDescPath_ = "/contrib/mcm-emulator/config/compilation/emulator_kmb_SC-Prefetch1.json";

template <typename T1, typename T2>
std::vector<T1> read(const std::string& filepath)
{
    std::ifstream fileStream(filepath, std::ifstream::binary);
    if (!fileStream) return {};
    std::vector<T1> data;
    T2 aux;
    while (fileStream.read(&reinterpret_cast<char&>(aux), sizeof(aux))) data.emplace_back(aux);
    return data;
}

template <typename T1, typename T2>
void write(const std::vector<T1>& data, const std::string& filepath)
{
    std::ofstream file(filepath, std::ofstream::binary);
    T2 aux;
    for (const auto& value: data)
    {
        aux = value;
        file.write(&reinterpret_cast<char&>(aux), sizeof(aux));
    };
}

mv::CompilationUnit::CompilationUnit(const std::string& modelName) :
model_(new OpModel(modelName))
{
    MV_PROFILER_START;
}

mv::CompilationUnit::~CompilationUnit()
{
    MV_PROFILER_FINISH("profiler_output.prof");
    delete model_;
}

bool mv::CompilationUnit::loadTargetDescriptor(const std::string& path)
{
    try
    {
        return targetDescriptor_.load(path);
    }
    catch (ArgumentError& e)
    {
        log(Logger::MessageType::Error, e.what());
        return false;
    }

    return true;

}

bool mv::CompilationUnit::loadCompilationDescriptor(const std::string& filePath)
{
    try
    {
        mv::json::Object jsonDesc = mv::CompilationDescriptor::load(filePath);
        compDescriptor_ = CompilationDescriptor(jsonDesc);
    }
    catch (ParsingError& e)
    {
        return false;
    }

    try 
    {   // parse recording on/off, but don't fail. Log message.
        if (compDescriptor_.getPassArg("initialize", "Singular", "GlobalConfigParams", "recorded_model"))
            model_->initRecordingFile("templateExampleNew.cpp");
    }
    catch (mv::AttributeError& e)
    {
        log(Logger::MessageType::Warning, "Could not find 'recorded_model' entry in 'GlobalConfigParams' section. Recording not enabled.");
    }
    return true;
}

bool mv::CompilationUnit::loadCompilationDescriptor(Target target)
{
    std::string descPath;

    switch (target)
    {
        case Target::ma2480:
        {
            descPath = utils::projectRootPath() + ma2480DefCompDescPath_;
            break;
        }
        case Target::ma2490:
        {
            descPath = utils::projectRootPath() + ma2490DefCompDescPath_;
            break;
        }
        default:
            return false;
    }

    return loadCompilationDescriptor(descPath);
}

bool mv::CompilationUnit::loadTargetDescriptor(Target target)
{

    switch (target)
    {

        case Target::ma2480:
        {
            std::string descPath = utils::projectRootPath() + ma2480DefTargetDescPath_;
            return loadTargetDescriptor(descPath);
        }

        case Target::ma2490:
        {
            std::string descPath = utils::projectRootPath() + ma2490DefTargetDescPath_;
            return loadTargetDescriptor(descPath);
        }

        default:
            return false;

    }

    return false;

}

mv::CompilationDescriptor& mv::CompilationUnit::compilationDescriptor()
{
    return compDescriptor_;
}

mv::OpModel& mv::CompilationUnit::model()
{
    return *model_;
}

bool mv::CompilationUnit::initialize()
{

    if (!passManager_.initialize(*model_, targetDescriptor_, compDescriptor_))
        return false;
    // Initialize resouces
    for (auto it = targetDescriptor_.memoryDefs().begin(); it != targetDescriptor_.memoryDefs().end(); ++it)
    {
        mv::DataModel dm(*model_);
        dm.addAllocator(it->first, it->second.size, it->second.alignment);
    }

    return true;

}

mv::Element mv::CompilationUnit::runStep()
{
    return passManager_.step();
}

mv::Element mv::CompilationUnit::run()
{
    MV_PROFILED_FUNCTION(MV_PROFILE_PHASE);
    Element output("CompilationOutput");
    output.set<std::string>("ModelName", model_->getName());
    std::vector<mv::Element> passList = compDescriptor_.serializePassList();
    passManager_.loadPassList(passList);

    // generate emulator results
    mv::Element globalParams = passList[0];
    bool emulator=false;
    if (globalParams.hasAttr("emulator_results") )
        emulator = globalParams.get<bool>("emulator_results");
    
    if (emulator)
        generateExpectedResults();

    while (!passManager_.completed())
    {
        MV_PROFILE_VIRTUAL_MEM;
        MV_PROFILE_PHYSICAL_MEM;
        #ifdef MV_PROFILER_ENABLED
        std::size_t ops = model_->opsCount();
        std::size_t dataFlows = model_->dataFlowsCount();
        ControlModel cm(*model_);
        std::size_t controlFlows = cm.controlFlowsCount();
        DataModel dm(*model_);
        std::size_t tensors = dm.tensorsCount();
        std::size_t populatedSize = dm.populatedTotalSize();
        std::size_t unpopulatedSize = dm.unpopulatedTotalSize();
        MV_PROFILED_VARIABLE(ops, MV_PROFILER_COLOR_GREEN);
        MV_PROFILED_VARIABLE(dataFlows, MV_PROFILER_COLOR_ROSE);
        MV_PROFILED_VARIABLE(controlFlows, MV_PROFILER_COLOR_RED);
        MV_PROFILED_VARIABLE(tensors, MV_PROFILER_COLOR_BLUE);
        MV_PROFILED_VARIABLE(populatedSize, MV_PROFILER_COLOR_ORANGE);
        MV_PROFILED_VARIABLE(unpopulatedSize, MV_PROFILER_COLOR_LIME);
        #endif

        output = passManager_.step();
    }

    return output;
}

bool mv::CompilationUnit::completed() const
{
    return passManager_.completed();
}

std::string mv::CompilationUnit::getLogID() const
{
    return "CompilationUnit";
}

std::shared_ptr<std::vector<char>> mv::CompilationUnit::getBlob() const
{
    if(!completed())
        log(Logger::MessageType::Warning, "Getting a blob from compilation unit before completion");
    mv::RuntimeModel& rm = mv::RuntimeModel::getInstance(targetDescriptor_);
    return rm.getBlob();
}

/**
 * Generates a deep copy of the opmodel
 */
void mv::CompilationUnit::deepCopy(mv::OpModel& copyTo)
{
    // std::cout << "Original Model:" << std::endl;
    DataModel dm(*model_);
    // ControlModel cm(*model_);
    // std::cout << "OpsCount: " << model_->opsCount() << std::endl;
    // std::cout << "dataFlows: " << model_->dataFlowsCount() << std::endl;    
    // std::cout << "controlFlows: " << cm.controlFlowsCount() << std::endl;
    // std::cout << "tensorsCount: " << dm.tensorsCount() << std::endl;
    // std::cout << "populatedSize: " << dm.populatedTotalSize() << std::endl;
    // std::cout << "unpopulatedSize: " << dm.unpopulatedTotalSize() << std::endl;

    for(auto opIterator = model_->opBegin(); opIterator != model_->opEnd(); ++opIterator)
    {
        // getAttrs() returns map, defineOp() requires vector
        std::vector<std::pair<std::string, mv::Attribute>> vectAttrs;
        for (const auto &attr : opIterator->getAttrs())
            vectAttrs.push_back(attr);

        copyTo.defineOp(opIterator->getOpType(), opIterator->getInputTensor(), vectAttrs, opIterator->getName(), false, false);
    }

    // flow attribute is not being set for each operation in copied model. Manually adding...
    for (Data::FlowListIterator flow = dm.flowBegin(); flow != dm.flowEnd(); ++flow)
    {
        std::string tensorName = flow->getTensor()->getName();
        mv::Data::TensorIterator newTensor = copyTo.getTensor(tensorName);

        if(!newTensor->hasAttr("flows"))
        {
            std::set<std::string> toSet;
            newTensor->set<std::set<std::string>>("flows", toSet);
        }
        newTensor->get<std::set<std::string>>("flows").insert(flow->getName());
    }
}

void mv::CompilationUnit::generateExpectedResults()
{   
    log(mv::Logger::MessageType::Debug, "Initializing emulator...");
    std::cout << "Initializing emulator..." << std::endl;
    mv::CompilationUnit emUnit(model_->getName());
    mv::OpModel& emOM = emUnit.model();
    deepCopy(emOM);

    // std::cout << std::endl << "Copied:" << std::endl;
    // std::cout << "OpsCount: " << emOM.opsCount() << std::endl;
    // std::cout << "dataFlows: " << emOM.dataFlowsCount() << std::endl;
    // ControlModel cm1(emOM);
    // std::cout << "controlFlows: " << cm1.controlFlowsCount() << std::endl;
    // DataModel dm1(emOM);
    // std::cout << "tensorsCount: " << dm1.tensorsCount() << std::endl;
    // std::cout << "populatedSize: " << dm1.populatedTotalSize() << std::endl;
    // std::cout << "unpopulatedSize: " << dm1.unpopulatedTotalSize() << std::endl;

    emUnit.loadTargetDescriptor(mv::Target::ma2490);
    std::string emuCompPath = utils::projectRootPath() + ma2490EmulatorCompDescPath_;
    std::cout <<  "loading Comp desc: " << emuCompPath << std::endl;
    emUnit.loadCompilationDescriptor(emuCompPath);
    emUnit.compilationDescriptor().setPassArg("GlobalConfigParams", "emulator_results", false); // prevent infinite loop

    emUnit.initialize();
    emUnit.run();
    std::cout << "Compilation completed..." << std::endl;

    // initialize the Emulator Manager
    mv::emu::Manager emulatorManager(emOM);

    // set input tensor values
    std::vector<std::int64_t> input0Data = read<std::int64_t, std::uint8_t>("./input.dat");
    if (input0Data.empty() ) throw RuntimeError(*this, "Emulator required file 'input.dat' not found in current directory");

    mv::Data::OpListIterator opIt = emOM.getOps("Input")[0];
    mv::Data::TensorIterator tensorIt = emOM.getOps("Input")[0]->getOutputTensor()[0];
    
    emulatorManager.populate(*(emOM.getOps("Input")[0]->getOutputTensor()[0]), mv::Order::getZMajorID(4), input0Data);
    
    std::cout << "Generating results..." << std::endl;
    emulatorManager.run();
}