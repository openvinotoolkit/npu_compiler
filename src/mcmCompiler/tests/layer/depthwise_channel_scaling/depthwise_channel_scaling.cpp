#include "include/mcm/compiler/compilation_unit.hpp"
#include <iostream>
#include <fstream>

int main()
{
    double inf = std::numeric_limits<double>::infinity();

    mv::CompilationUnit unit("DepthwiseChannelScalingModel");
    mv::OpModel& om = unit.model();
    auto input0 = om.input("input#3", {16,16,16,1}, mv::DType("UInt8"), mv::Order::getZMajorID(4));
    input0->setQuantParams({{128},{0.007843137718737125},{-1.0},{1.0}});

    std::vector<double> scale(16);
    for (std::size_t i = 0; i < 16; i++)
    {
        if (i%2==0)
        {
            scale[i] = 0.0015868041664361954;
        }
        else
        {
            scale[i] = 0.0015868041664361954/2;
        }
    }

    // Load weights from file
    std::vector<int64_t> d_weightsData0(1*1*16*1);
    std::string Weights_filename(mv::utils::projectRootPath() + "/tests/layer/depthwise_channel_scaling/depthwise_channel_scaling.in2");
    std::ifstream w_file;
    w_file.open(Weights_filename, std::fstream::in | std::fstream::binary);
    w_file.read((char*)(d_weightsData0.data()), 16 * sizeof(uint64_t));

    // Load biases from file
    std::vector<int64_t> biasd_WeightsData0(16);
    std::string Biases_filename(mv::utils::projectRootPath() + "/tests/layer/depthwise_channel_scaling/depthwise_channel_scaling.in3");
    std::ifstream b_file;
    b_file.open(Biases_filename, std::fstream::in | std::fstream::binary);
    b_file.read((char*)(biasd_WeightsData0.data()), 16 * sizeof(uint64_t));

    auto d_weights0 = om.constantInt("dwconv0#0_weights#1", d_weightsData0,{1,1,16,1}, mv::DType("UInt8"), mv::Order::getZMajorID(4));
    auto depthConv0 = om.depthwiseConv("dwconv0#4", input0, d_weights0, {1, 1}, {0, 0, 0, 0}, 1);
    d_weights0->setQuantParams({{145},{scale},{-0.2301538735628128},{0.17448118329048157}});
    depthConv0->setQuantParams({{128},{0.007843137718737125},{-1.003921627998352},{0.9960784316062927}});

    auto biasdWeights0 = om.constantInt("dwconv0#0_bias#2", biasd_WeightsData0,{16}, mv::DType("Int32"), mv::Order::getColMajorID(1));
    auto bias_cd0 = om.bias("", depthConv0, biasdWeights0);
    biasdWeights0->setQuantParams({{0},{1.2445522770576645e-05},{-inf},{inf}});
    bias_cd0->setQuantParams({{128},{0.007843137718737125},{-1.003921627998352},{0.9960784316062927}});

    om.output("", bias_cd0);

    std::string compDescPath = mv::utils::projectRootPath() + "/config/compilation/release_kmb.json";
    unit.loadCompilationDescriptor(compDescPath);
    unit.loadTargetDescriptor(mv::Target::ma2490);
    unit.initialize();
    unit.run();

    return 0;
}
