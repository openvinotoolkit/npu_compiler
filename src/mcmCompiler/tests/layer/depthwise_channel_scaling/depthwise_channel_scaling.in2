version https://git-lfs.github.com/spec/v1
oid sha256:f787cd3010227f2119837b7324a1645233a40d6798317e2fe27c09b4271ba71a
size 128
