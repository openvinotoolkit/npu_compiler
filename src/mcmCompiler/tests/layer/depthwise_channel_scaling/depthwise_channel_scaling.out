version https://git-lfs.github.com/spec/v1
oid sha256:69575566b471b8c13fed676703b1418f42e6576bb2b58f14bac56cd035a01129
size 4096
